samples/kinesis_file_gstreamer_sample_app.cpp

```diff
@@ -0,0 +1,797 @@
+#include <gst/gst.h>
+#include <gst/app/gstappsink.h>
+#include <string.h>
+#include <chrono>
+#include <Logger.h>
+#include "KinesisVideoProducer.h"
+#include <vector>
+#include <stdlib.h>
+#include <mutex>
+#include <IotCertCredentialProvider.h>
+
+using namespace std;
+using namespace com::amazonaws::kinesis::video;
+using namespace log4cplus;
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+int gstreamer_init(int, char **);
+
+#ifdef __cplusplus
+}
+#endif
+
+LOGGER_TAG("com.amazonaws.kinesis.video.gstreamer");
+
+#define DEFAULT_RETENTION_PERIOD_HOURS 2
+#define DEFAULT_KMS_KEY_ID ""
+#define DEFAULT_STREAMING_TYPE STREAMING_TYPE_OFFLINE
+#define DEFAULT_CONTENT_TYPE "video/h264,audio/aac"
+#define DEFAULT_MAX_LATENCY_SECONDS 60
+#define DEFAULT_FRAGMENT_DURATION_MILLISECONDS 2000
+#define DEFAULT_TIMECODE_SCALE_MILLISECONDS 1
+#define DEFAULT_KEY_FRAME_FRAGMENTATION TRUE
+#define DEFAULT_FRAME_TIMECODES TRUE
+#define DEFAULT_ABSOLUTE_FRAGMENT_TIMES TRUE
+#define DEFAULT_FRAGMENT_ACKS TRUE
+#define DEFAULT_RESTART_ON_ERROR TRUE
+#define DEFAULT_RECALCULATE_METRICS TRUE
+#define DEFAULT_STREAM_FRAMERATE 25
+#define DEFAULT_AVG_BANDWIDTH_BPS (4 * 1024 * 1024)
+#define DEFAULT_BUFFER_DURATION_SECONDS 120
+#define DEFAULT_REPLAY_DURATION_SECONDS 40
+#define DEFAULT_CONNECTION_STALENESS_SECONDS 60
+#define DEFAULT_CODEC_ID "V_MPEG4/ISO/AVC"
+#define DEFAULT_TRACKNAME "kinesis_video"
+#define DEFAULT_FRAME_DURATION_MS 1
+#define DEFAULT_CREDENTIAL_ROTATION_SECONDS 3600
+#define DEFAULT_CREDENTIAL_EXPIRATION_SECONDS 180
+#define DEFAULT_TOTAL_TRACK_COUNT 2
+#define DEFAULT_AUDIO_VIDEO_DRIFT_TIMEOUT_SECOND 5
+
+#define DEFAULT_VIDEO_TRACKID 1
+#define DEFAULT_AUDIO_TRACK_NAME "audio"
+#define DEFAULT_AUDIO_CODEC_ID "A_AAC"
+#define DEFAULT_AUDIO_TRACKID 2
+
+typedef struct _FileInfo {
+    _FileInfo():
+            path(""),
+            last_fragment_ts(0) {}
+    string path;
+    uint64_t last_fragment_ts;
+    mutex mtx;
+} FileInfo;
+
+typedef struct _CustomData {
+
+    _CustomData():
+            first_video_frame(true),
+            eos_triggered(false),
+            stream_status(STATUS_SUCCESS),
+            key_frame_pts(0),
+            main_loop(NULL) {
+        producer_start_time = chrono::duration_cast<nanoseconds>(systemCurrentTime().time_since_epoch()).count();
+    }
+
+    GMainLoop *main_loop;
+    // queue elements to be linked to demuxer's sometimes pad in demux_pad_cb
+    GstElement *video_queue, *audio_queue;
+    GstElement *pipeline;
+
+    unique_ptr<KinesisVideoProducer> kinesis_video_producer;
+    shared_ptr<KinesisVideoStream> kinesis_video_stream;
+    char *stream_name;
+
+    mutex audio_video_sync_mtx;
+    condition_variable audio_video_sync_cv;
+
+    // file to upload
+    FileInfo file;
+
+    // indicate whether a video key frame has been received or not.
+    volatile bool first_video_frame;
```

Comment:
Might want atomic_bool here too

Reply:
No, it's alright since it's already protected by the mutex, https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/pull/442/files#diff-2af4741aceb98d75d1fb60eac19ef4f3R316